# <a id='toc1_'></a>[Grouping Data & Aggregating Variables](#toc0_)

**Table of contents**<a id='toc0_'></a>    
- [Grouping Data & Aggregating Variables](#toc1_)    
  - [I. Example Data Aggregration](#toc1_1_)    
    - [I.1. Data aggregation with the agg() function](#toc1_1_1_)    
      - [I.1.1. Group data by column 'department_id'](#toc1_1_1_1_)    
      - [I.1.2. Mean, min and max of 'order_number'](#toc1_1_1_2_)    
    - [I.2. Data aggregation with the transform() function](#toc1_1_2_)    
      - [I.2.1. Create a new column 'max_order' to place the results of the maximum orders for each user](#toc1_1_2_1_)    
  - [II. Data Aggregation for Exploratory Data Analysis](#toc1_2_)    
    - [II.1. Create a flag that assigns a loyalty label to a user ID based on its corresponding max order value](#toc1_2_1_)    
    - [II.2. Determine the spending habits of each loyalty category in column 'loyalty_flag'](#toc1_2_2_)    
      - [II.2.1. Data consistency check](#toc1_2_2_1_)    
        - [II.2.1.1. Check for abnormal values](#toc1_2_2_1_1_)    
        - [II.2.1.2. Address abnormal values](#toc1_2_2_1_2_)    
    - [II.3. Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers](#toc1_2_3_)    
        - [Observations:](#toc1_2_3_1_1_)    
    - [II.4. Create a spending flag for each user based on the average price across all their orders](#toc1_2_4_)    
    - [II.5. Create an order frequency flag that marks the regularity of a user’s ordering behavior](#toc1_2_5_)    
      - [II.5.1. Create new column 'order_freq'](#toc1_2_5_1_)    
      - [II.5.2. Data consistency check & data wrangling](#toc1_2_5_2_)    
        - [II.5.2.1. Check for null values](#toc1_2_5_2_1_)    
        - [II.5.2.2. Rename columns](#toc1_2_5_2_2_)    
  - [III. Data Export](#toc1_3_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# create a path to the directory
path = r'C:\Users\Ansgar.S\Uyen\OneDrive\Documents\Data Immersion\Achievement IV - Python Fundamentals for Data Analysts\02-2023 Instacart Basket Analysis'

# import the 'orders_products_merged_2.pkl' dataset
df_ords_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_2.pkl'))

## <a id='toc1_1_'></a>[I. Example Data Aggregration](#toc0_)

### <a id='toc1_1_1_'></a>[I.1. Data aggregation with the agg() function](#toc0_)

#### <a id='toc1_1_1_1_'></a>[I.1.1. Group data by column 'department_id'](#toc0_)

In [3]:
# create a filter of df_ords_merged
df = df_ords_merged[:1000000]

print('Sample output of filter of df_ords_merged:')
df.sample(5)

Sample output of filter of df_ords_merged:


product_id                      product_name aisle_id department_id  \
936929       1942           Aged White Wine Vinegar       19            13   
851657       1729              2% Lactose Free Milk       84            16   
164713        260                        Cantaloupe       24             4   
746792       1463                      Organic Milk       84            16   
869544       1804  Baked White Cheddar Cheese Puffs      107            19   

        prices order_id user_id  order_number  orders_day_of_week  \
936929     4.2  1761870   20179            13                   0   
851657    11.6    61515  180719            74                   2   
164713     1.4  1752672   71027            15                   0   
746792    11.3  1988609  176793             1                   2   
869544     3.6   656392  167276             9                   6   

        order_hour_of_day  days_since_prior_order  is_new_customer  \
936929                 10                     5.0            False   
851657                 14                     2.0            False   
164713                 15                     7.0            False   
746792                  8                     NaN             True   
869544                 15                     7.0            False   

        add_to_cart_order  reordered        price_range     busiest_day  \
936929                 10          0  Low-range product     Busiest day   
851657                  2          1  Mid-range product  Regularly busy   
164713                 15          1  Low-range product     Busiest day   
746792                  3          0  Mid-range product  Regularly busy   
869544                  6          1  Low-range product  Regularly busy   

        busiest_days busiest_period_of_day  
936929  Busiest days           Most orders  
851657  Regular days           Most orders  
164713  Busiest days           Most orders  
746792  Regular days        Average orders  
869544  Regular days           Most orders

#### <a id='toc1_1_1_2_'></a>[I.1.2. Mean, min and max of 'order_number'](#toc0_)

In [4]:
# group dataframe df by column 'department_id', then calculate the mean, min and max of column 'order_number'
print('Mean, min and max of order_number grouped by column department_id in df:')
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

Mean, min and max of order_number grouped by column department_id in df:


order_number        
                      mean min max
department_id                     
1                15.577493   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
2                17.320781   1  96
20               16.255442   1  99
21               25.535479   1  98
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99

In [5]:
# check output of df_ords_merged
print('Sample output of of df_ords_merged:')
df_ords_merged.sample(5)

Sample output of of df_ords_merged:


product_id                   product_name aisle_id department_id  \
32087091      49235            Organic Half & Half       53            16   
19427667      29700                Soft Toothbrush       20            11   
21811322      33719  French Vanilla Yogurt Vanilla      120            16   
22699745      35108                  Salted Butter       36            16   
13853508      22227              Coconut Cream Pie        3            19   

          prices order_id user_id  order_number  orders_day_of_week  \
32087091     1.8  2975766  146375             8                   4   
19427667     8.5  2759152   49064            35                   0   
21811322     4.3  3417956  137834            43                   1   
22699745     6.5  1006872   52265             3                   3   
13853508     2.2   708522   42019             3                   6   

          order_hour_of_day  days_since_prior_order  is_new_customer  \
32087091                 12                    12.0            False   
19427667                  8                     0.0            False   
21811322                 17                     2.0            False   
22699745                 10                     8.0            False   
13853508                 16                     8.0            False   

          add_to_cart_order  reordered        price_range     busiest_day  \
32087091                  7          0  Low-range product      Least busy   
19427667                 11          1  Mid-range product     Busiest day   
21811322                  2          1  Low-range product  Regularly busy   
22699745                 16          0  Mid-range product  Regularly busy   
13853508                  9          0  Low-range product  Regularly busy   

          busiest_days busiest_period_of_day  
32087091  Slowest days           Most orders  
19427667  Busiest days        Average orders  
21811322  Busiest days        Average orders  
22699745  Slowest days           Most orders  
13853508  Regular days           Most orders

In [6]:
print('Number of rows and columns in df_ords_merged:')
df_ords_merged.shape

Number of rows and columns in df_ords_merged:


(32404859, 18)

In [7]:
# group dataframe df_ords_merged by column 'department_id', then calculate the mean, min and max of column 'order_number'
print('Mean, min and max of order_number grouped by column department_id in df_ords_merged:')
df_ords_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

Mean, min and max of order_number grouped by column department_id in df_ords_merged:


order_number        
                      mean min max
department_id                     
1                15.457838   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
2                17.277920   1  99
20               16.473447   1  99
21               22.902379   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99

**Observation**:
The aggregations in df and df_ords_merged are very similar to each other, with some minor differences in mean and max values.

### <a id='toc1_1_2_'></a>[I.2. Data aggregation with the transform() function](#toc0_)

#### <a id='toc1_1_2_1_'></a>[I.2.1. Create a new column 'max_order' to place the results of the maximum orders for each user](#toc0_)

In [8]:
# create a new column 'max_order' containing the maximum order number for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['max_order'] = df_ords_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
# check the output of 'user_id' and 'max_order'
print('Sample user_id and max_order output:')
df_ords_merged[['user_id', 'max_order']].sample(10)

Sample user_id and max_order output:


user_id  max_order
27177657   81583         41
12807098    3411         19
11912791   45446         20
14192919  146369         72
2679246    29969         34
23665713  174791          5
24794509   31903         38
8076674   116698         89
18090472  122354         19
15223682   94556         42

## <a id='toc1_2_'></a>[II. Data Aggregation for Exploratory Data Analysis](#toc0_)

### <a id='toc1_2_1_'></a>[II.1. Create a flag that assigns a loyalty label to a user ID based on its corresponding max order value](#toc0_)

In [10]:
# create column 'loyalty_flag' based on each user's maximum order value
df_ords_merged.loc[df_ords_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_merged.loc[(df_ords_merged['max_order'] <= 40) & (df_ords_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_merged.loc[df_ords_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
# count the values in column 'loyalty_flag'
print('The value counts in column loyalty_flag:')
df_ords_merged['loyalty_flag'].value_counts(dropna = False)

The value counts in column loyalty_flag:


Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [12]:
# check the accuracy of column 'loyalty_flag' with output from columns 'user_id', 'max_order', and 'loyalty_flag'
print('Sample outputs of columns user_id, max_order and loyalty_flag:')
df_ords_merged[['user_id', 'max_order', 'loyalty_flag']].sample(20)

Sample outputs of columns user_id, max_order and loyalty_flag:


user_id  max_order      loyalty_flag
20033621  203411         30  Regular customer
19691998   33813         29  Regular customer
9784409    99555         12  Regular customer
2660331     4090         53    Loyal customer
11843685  149033         56    Loyal customer
24586632  133708         26  Regular customer
22900499  140849         21  Regular customer
7591349    39600          9      New customer
14717535   83326         99    Loyal customer
19361148  196994         14  Regular customer
21974688  101411          6      New customer
20335938  138798         18  Regular customer
30390699  141277         29  Regular customer
24229516   77264         76    Loyal customer
489424      2888          6      New customer
17914639  155324         10      New customer
16766034   67626         40  Regular customer
1988921   175854         35  Regular customer
11623363   15194         13  Regular customer
15141449  114084         68    Loyal customer

### <a id='toc1_2_2_'></a>[II.2. Determine the spending habits of each loyalty category in column 'loyalty_flag'](#toc0_)

In [13]:
# basic statistics of df_ords_merged by column 'loyalty_flag', then calculate the mean, min and max of column 'prices'
print('Mean, min and max of prices grouped by column loyalty_flag in df_ords_merged:')
df_ords_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

Mean, min and max of prices grouped by column loyalty_flag in df_ords_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

**Observation**:
On average, new customers go for the highest prices, followed by regular customers and surprisingly, loyal customers buy the cheapest products. However, the maximum price is 99,999 has skewed the distribution. Therefore, statistics need to be considered when this value is removed from the dataframe.

In [14]:
# run descriptive statistics on column 'prices'
print('Descriptive statistics of column prices:')
df_ords_merged['prices'].describe()

Descriptive statistics of column prices:


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

**Observation**:
The most unusual observation from these statistics is the large standard deviation of 495,655. This indicates that there is a large amount of variation in the data, with some values being much higher than the mean of 11.8. The maximum value of 99,999 is also much higher than the mean, indicating that there are some very large values in the column.

#### <a id='toc1_2_2_1_'></a>[II.2.1. Data consistency check](#toc0_)

##### <a id='toc1_2_2_1_1_'></a>[II.2.1.1. Check for abnormal values](#toc0_)

In [15]:
# check for abnormal values in column 'prices'
print('Check for abnormal values in column prices:')
df_ords_merged['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in column prices:


21786183    99999.0
13100147    14900.0
5254814        25.0
5784936        24.9
5427379        24.8
             ...   
156617          1.4
11863           1.3
66073           1.2
2601            1.1
146043          1.0
Name: prices, Length: 242, dtype: float64

**Observation**:
The prices 99,999 and 14,900 seem unusually high.

In [16]:
# get the output of df_ords_merged when the value in column 'prices' is 99999 or 14900
abnormal_prices = df_ords_merged.loc[df_ords_merged['prices'].isin([99999, 14900])]

In [17]:
print('Sample outputs of abnormal_prices:')
abnormal_prices.sample(5)

Sample outputs of abnormal_prices:


product_id                      product_name aisle_id department_id  \
21786444      33664             2 % Reduced Fat  Milk       84            16   
13100296      21553  Lowfat 2% Milkfat Cottage Cheese      108            16   
13104010      21553  Lowfat 2% Milkfat Cottage Cheese      108            16   
13102405      21553  Lowfat 2% Milkfat Cottage Cheese      108            16   
21786739      33664             2 % Reduced Fat  Milk       84            16   

           prices order_id user_id  order_number  orders_day_of_week  \
21786444  99999.0  1429377   93841            10                   0   
13100296  14900.0  3088963    7151             7                   6   
13104010  14900.0  3222303  180844             2                   4   
13102405  14900.0  2170182  105483             5                   3   
21786739  99999.0   445923  164929            37                   2   

          order_hour_of_day  days_since_prior_order  is_new_customer  \
21786444                 17                     5.0            False   
13100296                 14                     8.0            False   
13104010                 14                    15.0            False   
13102405                 13                    30.0            False   
21786739                 14                    12.0            False   

          add_to_cart_order  reordered         price_range     busiest_day  \
21786444                 10          1  High-range product     Busiest day   
13100296                  6          1  High-range product  Regularly busy   
13104010                 15          1  High-range product      Least busy   
13102405                  5          1  High-range product  Regularly busy   
21786739                  3          1  High-range product  Regularly busy   

          busiest_days busiest_period_of_day  max_order      loyalty_flag  
21786444  Busiest days        Average orders         13  Regular customer  
13100296  Regular days           Most orders         41    Loyal customer  
13104010  Slowest days           Most orders          6      New customer  
13102405  Slowest days           Most orders          8      New customer  
21786739  Regular days           Most orders         37  Regular customer

In [18]:
# descriptive statistics on abnormal_prices
print('Descriptive statistics on abnormal_prices:')
abnormal_prices.describe()

Descriptive statistics on abnormal_prices:


prices  order_number  orders_day_of_week  order_hour_of_day  \
count   5127.000000   5127.000000         5127.000000        5127.000000   
mean   26485.547494     15.302126            2.750536          13.473376   
std    29186.636448     16.691849            2.010892           4.137326   
min    14900.000000      1.000000            0.000000           0.000000   
25%    14900.000000      4.000000            1.000000          10.000000   
50%    14900.000000      9.000000            3.000000          13.000000   
75%    14900.000000     20.000000            5.000000          16.000000   
max    99999.000000     99.000000            6.000000          23.000000   

       days_since_prior_order  add_to_cart_order    reordered    max_order  
count             4776.000000        5127.000000  5127.000000  5127.000000  
mean                11.899079           7.986347     0.613614    29.791691  
std                  9.068019           6.868889     0.486968    24.929207  
min                  0.000000           1.000000     0.000000     3.000000  
25%                  5.000000           3.000000     0.000000    12.000000  
50%                  8.000000           6.000000     1.000000    21.000000  
75%                 16.000000          11.000000     1.000000    41.000000  
max                 30.000000          72.000000     1.000000    99.000000

**Observation**:
The unusually high prices are both from the department ID 16. The pricetag 14,900 belongs to the product ID 21553, the 'Lowfat 2% Milkfat Cottage Cheese', while the pricetag 99,999 belongs to the product ID 33664, named '2 % Reduced Fat Milk'.

##### <a id='toc1_2_2_1_2_'></a>[II.2.1.2. Address abnormal values](#toc0_)

In [19]:
# create a filter 'dairy' for records with the department_id value of 16
dairy = df_ords_merged.loc[df_ords_merged['department_id'].isin(['16'])]

In [20]:
# create a filter 'dairy_without_weird' to exclude records in dairy without the unusually high prices
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]

In [21]:
print('Sampled outputs of dairy_without_weird:')
dairy_without_weird.sample(10)

Sampled outputs of dairy_without_weird:


product_id                              product_name aisle_id  \
14405664      23044  Cream Top Smooth & Creamy Vanilla Yogurt      120   
30149944      46886                         Raspberry Yoghurt      120   
30803976      47611               M&M's Vanilla Lowfat Yogurt      120   
2054444        4274         Vegan Cheddar Style Cheese Slices       21   
14859827      23909                       2% Reduced Fat Milk       84   
19828173      30442                    Low Fat Vanilla Yogurt      120   
32061220      49235                       Organic Half & Half       53   
16106113      25133                     Organic String Cheese       21   
6943969       12099                        Honey Greek Yogurt      120   
19334409      29544      Cream Top Peach on the Bottom Yogurt      120   

         department_id  prices order_id user_id  order_number  \
14405664            16    12.9  2132183   36171            30   
30149944            16    13.2  2772720  108089            34   
30803976            16    14.8  3082671  101319             9   
2054444             16     8.4  1831547  122254            16   
14859827            16     9.2  1809330    6990             1   
19828173            16    10.5  3073445  136336             9   
32061220            16     1.8  3096158   76484            18   
16106113            16     8.6  1231151   32785            20   
6943969             16     4.1  2275730  162135            10   
19334409            16    11.4  1493572  111774            55   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
14405664                   4                 12                     7.0   
30149944                   4                  6                    11.0   
30803976                   3                 21                     1.0   
2054444                    1                 17                    12.0   
14859827                   1                 18                     NaN   
19828173                   4                 12                     7.0   
32061220                   0                  8                    17.0   
16106113                   1                 21                     7.0   
6943969                    6                 10                     6.0   
19334409                   2                 18                     4.0   

          is_new_customer  add_to_cart_order  reordered        price_range  \
14405664            False                  9          1  Mid-range product   
30149944            False                 36          0  Mid-range product   
30803976            False                 12          0  Mid-range product   
2054444             False                  2          1  Mid-range product   
14859827             True                 10          0  Mid-range product   
19828173            False                  9          0  Mid-range product   
32061220            False                  9          1  Low-range product   
16106113            False                  6          1  Mid-range product   
6943969             False                  3          1  Low-range product   
19334409            False                  2          1  Mid-range product   

             busiest_day  busiest_days busiest_period_of_day  max_order  \
14405664      Least busy  Slowest days           Most orders         42   
30149944      Least busy  Slowest days         Fewest orders         38   
30803976  Regularly busy  Slowest days        Average orders         99   
2054444   Regularly busy  Busiest days        Average orders         27   
14859827  Regularly busy  Busiest days        Average orders         28   
19828173      Least busy  Slowest days           Most orders         10   
32061220     Busiest day  Busiest days        Average orders         23   
16106113  Regularly busy  Busiest days        Average orders         37   
6943969   Regularly busy  Regular days           Most orders         49   
19334409  Regularly busy  Regular days     

In [22]:
# check to see the filter is working properly
print('Maximum price in dairy_without_weird:')
dairy_without_weird['prices'].max()

Maximum price in dairy_without_weird:


15.0

In [23]:
# create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird
reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [24]:
print('Sample outputs of reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample outputs of reduced_fat_milk:


product_id                   product_name aisle_id department_id  \
927496         1940    Organic 2% Reduced Fat Milk       84            16   
25794529      40174    2% Reduced Fat Organic Milk       84            16   
14879318      23909            2% Reduced Fat Milk       84            16   
25809413      40174    2% Reduced Fat Organic Milk       84            16   
3292428        5785    Organic Reduced Fat 2% Milk       84            16   
3288335        5785    Organic Reduced Fat 2% Milk       84            16   
1432675        2962  Milk, Reduced Fat, 2% Milkfat       84            16   
21076553      32478            Reduced Fat 2% Milk       84            16   
14880755      23909            2% Reduced Fat Milk       84            16   
14881145      23909            2% Reduced Fat Milk       84            16   

          prices order_id user_id  order_number  orders_day_of_week  \
927496       9.1  1812391  122621             3                   4   
25794529    11.4  2604531    9657            20                   1   
14879318     9.2  2042777  117715             1                   5   
25809413    11.4  2301417  198480             7                   2   
3292428      7.9   208424  122853            17                   1   
3288335      7.9  2112069  106421            21                   4   
1432675      2.7  2831756   22297             3                   4   
21076553    13.9  2945273  201685            25                   6   
14880755     9.2  3003129  127528             1                   6   
14881145     9.2  2480775  129647            23                   3   

          order_hour_of_day  days_since_prior_order  is_new_customer  \
927496                   17                     3.0            False   
25794529                  7                     5.0            False   
14879318                 18                     NaN             True   
25809413                 11                     8.0            False   
3292428                  14                     5.0            False   
3288335                  15                     3.0            False   
1432675                  17                     5.0            False   
21076553                  0                    30.0            False   
14880755                  9                     NaN             True   
14881145                 11                     1.0            False   

          add_to_cart_order  reordered        price_range     busiest_day  \
927496                    2          0  Mid-range product      Least busy   
25794529                 10          1  Mid-range product  Regularly busy   
14879318                  2          0  Mid-range product  Regularly busy   
25809413                  3          1  Mid-range product  Regularly busy   
3292428                   3          1  Mid-range product  Regularly busy   
3288335                   8          1  Mid-range product      Least busy   
1432675                   9          1  Low-range product      Least busy   
21076553                  2          1  Mid-range product  Regularly busy   
14880755                  5          0  Mid-range product  Regularly busy   
14881145                  2          1  Mid-range product  Regularly busy   

          busiest_days busiest_period_of_day  max_order      loyalty_flag  
927496    Slowest days        Average orders         68    Loyal customer  
25794529  Busiest days        Average orders         44    Loyal customer  
14879318  Regular days        Average orders          6      New customer  
25809413  Regular days           Most orders          8      New customer  
3292428   Busiest days           Most orders         30  Regular customer  
3288335   Slowest days           Most orders         59    Loyal customer  
1432675   Slowest days        Average orders         14  Regular customer  
21076553  Regular days         Fewest orders         31  Regular customer  
14880755  Regular days           Most orders         2

In [25]:
# quick statistics of reduced_fat_milk
print('Quick statistics of reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics of reduced_fat_milk:


prices   order_number  orders_day_of_week  order_hour_of_day  \
count  202238.000000  202238.000000       202238.000000      202238.000000   
mean        9.086900      18.528684            2.748702          13.181835   
std         3.064869      18.288088            2.022770           4.193794   
min         1.200000       1.000000            0.000000           0.000000   
25%         7.900000       5.000000            1.000000          10.000000   
50%         9.200000      12.000000            3.000000          13.000000   
75%        11.400000      26.000000            5.000000          16.000000   
max        14.700000      99.000000            6.000000          23.000000   

       days_since_prior_order  add_to_cart_order      reordered      max_order  
count           190556.000000      202238.000000  202238.000000  202238.000000  
mean                10.538445           5.423051       0.782692      35.834769  
std                  8.380235           5.397505       0.412415      26.069071  
min                  0.000000           1.000000       0.000000       3.000000  
25%                  5.000000           2.000000       1.000000      14.000000  
50%                  7.000000           4.000000       1.000000      30.000000  
75%                 14.000000           7.000000       1.000000      51.000000  
max                 30.000000         137.000000       1.000000      99.000000

**Observation**:
The mean, min and max prices for products that contain 'Reduced Fat', '2' and 'Milk' in the name are: 9.01, 1.2, 14.7. As a result of the rationale, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99,999 to 9.99.

In [26]:
# create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird
cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [27]:
print('Sample outputs of cottage_cheese:')
cottage_cheese.sample(10)

Sample outputs of cottage_cheese:


product_id                                 product_name aisle_id  \
17982201      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17982685      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17981988      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17982537      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
3913462        6852   4% Milkfat Cottage Cheese Small Curd 22 OZ      108   
17982674      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17982534      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17982647      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17981478      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
17982389      27790  Small Curd Lowfat 2% Milkfat Cottage Cheese      108   

         department_id  prices order_id user_id  order_number  \
17982201            16     7.7  2495278   77686             6   
17982685            16     7.7   158385  110840            37   
17981988            16     7.7  1267460   57070            10   
17982537            16     7.7  1229820  103288            16   
3913462             16     8.6  2250055  198648             4   
17982674            16     7.7   368932  110840            16   
17982534            16     7.7   533454  103288             8   
17982647            16     7.7  1756518  110040            56   
17981478            16     7.7  2329996    8518            22   
17982389            16     7.7  1644862   94077             3   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
17982201                   5                 15                    30.0   
17982685                   6                  9                     9.0   
17981988                   2                 12                    14.0   
17982537                   5                 12                     7.0   
3913462                    3                  9                    30.0   
17982674                   5                  8                     2.0   
17982534                   5                 14                     7.0   
17982647                   1                 15                     0.0   
17981478                   3                 20                    13.0   
17982389                   0                  6                    30.0   

          is_new_customer  add_to_cart_order  reordered        price_range  \
17982201            False                  4          0  Mid-range product   
17982685            False                  1          1  Mid-range product   
17981988            False                  3          1  Mid-range product   
17982537            False                  5          1  Mid-range product   
3913462             False                  2          1  Mid-range product   
17982674            False                  1          1  Mid-range product   
17982534            False                  2          1  Mid-range product   
17982647            False                  8          1  Mid-range product   
17981478            False                  1          1  Mid-range product   
17982389            False                  5          1  Mid-range product   

             busiest_day  busiest_days busiest_period_of_day  max_order  \
17982201  Regularly busy  Regular days           Most orders         19   
17982685  Regularly busy  Regular days           Most orders         51   
17981988  Regularly busy  Regular days           Most orders         22   
17982537  Regularly busy  Regular days           Most orders         53   
3913462   Regularly busy  Slowest days           Most orders         14   
17982674  Regularly busy  Regular days        Average orders         51   
17982534  Regularly busy  Regular days           Most orders         53   
17982647  Regularly busy  Busiest days           Most orders         82   
17981478  Regularly busy  Slowest days        Average orders         30   
17982389  

In [28]:
# quick statistics in cottage_cheese
print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


prices  order_number  orders_day_of_week  order_hour_of_day  \
count  3487.000000   3487.000000         3487.000000        3487.000000   
mean      7.228219     15.484084            2.849728          13.257241   
std       1.757900     16.808776            2.031540           4.297401   
min       2.400000      1.000000            0.000000           0.000000   
25%       7.700000      4.000000            1.000000          10.000000   
50%       7.700000     10.000000            3.000000          13.000000   
75%       7.700000     20.000000            5.000000          16.000000   
max      11.800000     98.000000            6.000000          23.000000   

       days_since_prior_order  add_to_cart_order    reordered    max_order  
count             3218.000000        3487.000000  3487.000000  3487.000000  
mean                12.898073           5.840551     0.697448    30.531116  
std                  9.231128           5.955836     0.459429    25.413902  
min                  0.000000           1.000000     0.000000     3.000000  
25%                  6.000000           2.000000     0.000000    11.000000  
50%                 10.000000           4.000000     1.000000    22.000000  
75%                 19.000000           8.000000     1.000000    43.000000  
max                 30.000000          57.000000     1.000000    99.000000

**Observation**:
The mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name are: 7.23, 2.4, 11.8. As a result, the price of product ID 21553, the 'Lowfat 2% Milkfat Cottage Cheese' will be changed from 14,900, and imputed by the mean value mentioned.

In [29]:
# replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese
df_ords_merged['prices'] = df_ords_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [30]:
#check the statistics of column 'prices' in df_ords_merged
print('The statistics of column prices in df_ords_merged:')
df_ords_merged['prices'].describe()

The statistics of column prices in df_ords_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

### <a id='toc1_2_3_'></a>[II.3. Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers](#toc0_)

In [31]:
# basic statistics of df_ords_merged by 'loyalty_flag' with the updated 'prices' column, then calculate the mean, min and max of the 'prices' column
print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_ords_merged:')
df_ords_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_ords_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

##### <a id='toc1_2_3_1_1_'></a>[Observations:](#toc0_)
The spending ranking remains the same: new consumers pay the most for products, followed by regular customers, and finally loyal customers. The mean values, on the other hand, are considerably closer, and the maximum values have been lowered to 25.

### <a id='toc1_2_4_'></a>[II.4. Create a spending flag for each user based on the average price across all their orders](#toc0_)

In [32]:
# create a new column 'avg_price' containing the average price for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['avg_price'] = df_ords_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [33]:
# check the output of 'user_id' and 'avg_price'
print('Sample outputs of user_id and avg_price:')
df_ords_merged[['user_id', 'avg_price']].head(10)

Sample outputs of user_id and avg_price:


user_id  avg_price
0     138   6.935811
1     138   6.935811
2     709   7.930208
3     764   4.972414
4     764   4.972414
5     777   6.935398
6     825   5.957576
7     910   6.680000
8    1052   7.162500
9    1052   7.162500

**Create a flag with the criteria**:

If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [34]:
# create column 'spending_flag' based on each user's average price
df_ords_merged.loc[df_ords_merged['avg_price'] < 10, 'spending_flag'] = 'Low spender'
df_ords_merged.loc[df_ords_merged['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [35]:
# count the values in column 'spending_flag'
print('The value counts in column spending_flag:')
df_ords_merged['spending_flag'].value_counts(dropna = False)

The value counts in column spending_flag:


Low spender     32285150
High spender      119709
Name: spending_flag, dtype: int64

In [36]:
# check the accuracy of column 'spending_flag' with outputs from columns 'user_id', 'avg_price'
print('Sample outputs of columns user_id, avg_price:')
df_ords_merged[['user_id', 'avg_price', 'spending_flag']].sample(10)

Sample outputs of columns user_id, avg_price:


user_id  avg_price spending_flag
14368926   78686   6.971555   Low spender
3160880    70499   6.864045   Low spender
26385156   49860   7.824176   Low spender
2005035     4804   6.702604   Low spender
25715635     263   6.945833   Low spender
31878399  124192   7.216584   Low spender
28797770  150120   8.122881   Low spender
2638868   171932   9.925758   Low spender
4294180    53259   8.799578   Low spender
22678566    8652   7.298227   Low spender

### <a id='toc1_2_5_'></a>[II.5. Create an order frequency flag that marks the regularity of a user’s ordering behavior](#toc0_)

#### <a id='toc1_2_5_1_'></a>[II.5.1. Create new column 'order_freq'](#toc0_)

In [37]:
# create a new column 'order_freq' containing the median in days_since_prior_order for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['order_freq'] = df_ords_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [38]:
# check the output of 'user_id' and 'order_freq'
print('Sample outputs of user_id and order_freq:')
df_ords_merged[['user_id', 'order_freq']].head(10)

Sample outputs of user_id and order_freq:


user_id  order_freq
0     138         8.0
1     138         8.0
2     709         8.0
3     764         9.0
4     764         9.0
5     777        11.0
6     825        20.0
7     910         6.0
8    1052        10.0
9    1052        10.0

**Create a flag with the criteria**:

- If the median of 'days_since_prior_order' is higher than 20, then the customer should be labeled a 'Non-frequent customer'.

- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a 'Regular customer'.

- If the median is lower than or equal to 10, then the customer should be labeled a 'Frequent customer'.

In [39]:
# create the 'order_freq_flag' column based on each user's median in days_since_prior_order
df_ords_merged.loc[df_ords_merged['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'
df_ords_merged.loc[(df_ords_merged['order_freq'] > 10) & (df_ords_merged['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'
df_ords_merged.loc[df_ords_merged['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [40]:
# count the values in column 'order_freq_flag'
print('Value counts in column order_freq_flag:')
df_ords_merged['order_freq_flag'].value_counts(dropna = False)

Value counts in column order_freq_flag:


Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

#### <a id='toc1_2_5_2_'></a>[II.5.2. Data consistency check & data wrangling](#toc0_)

##### <a id='toc1_2_5_2_1_'></a>[II.5.2.1. Check for null values](#toc0_)

In [41]:
# check the null values
print('Records in df_ords_merged with null values in column order_freq_flag:')
df_ords_merged[df_ords_merged['order_freq_flag'].isnull()][['user_id', 'days_since_prior_order', 'order_freq', 'order_freq_flag']]

Records in df_ords_merged with null values in column order_freq_flag:


user_id  days_since_prior_order  order_freq order_freq_flag
6234909   159838                     NaN         NaN             NaN
12947653  159838                     NaN         NaN             NaN
13839012  159838                     NaN         NaN             NaN
14758536  159838                     NaN         NaN             NaN
21673807  159838                     NaN         NaN             NaN

In [42]:
# check records of user_id 159838
print('Records of user_id 159838:')
df_ords_merged[df_ords_merged['user_id'] == 159838]

Records of user_id 159838:


Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, is_new_customer, add_to_cart_order, reordered, price_range, busiest_day, busiest_days, busiest_period_of_day, max_order, loyalty_flag, avg_price, spending_flag, order_freq, order_freq_flag]
Index: []

[0 rows x 24 columns]

In [43]:
# check to see if user ID 159838 is a new customer
df_ords_merged[df_ords_merged['user_id'] == 159838]['is_new_customer']

Series([], Name: is_new_customer, dtype: bool)

**Observation**:
User ID 159838 seems to be a new customer, with only one purchase so far. The solution is to leave the records as is, since replacing the null values with 0 will change the order frequency flag into 'Frequent customer' instead.

In [44]:
# check the column names in df_ords_merged so far
print('Column names in df_ords_merged:')
df_ords_merged.columns

Column names in df_ords_merged:


Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'is_new_customer',
       'add_to_cart_order', 'reordered', 'price_range', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_price', 'spending_flag', 'order_freq', 'order_freq_flag'],
      dtype='object')

##### <a id='toc1_2_5_2_2_'></a>[II.5.2.2. Rename columns](#toc0_)

In [45]:
# change is_new_customer to a more intuitive name
df_ords_merged.rename(columns={'is_new_customer': 'first_purchase'}, inplace = True)

In [46]:
# check the column names in df_ords_merged again
print('Column names in df_ords_merged:')
df_ords_merged.columns

Column names in df_ords_merged:


Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'first_purchase',
       'add_to_cart_order', 'reordered', 'price_range', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_price', 'spending_flag', 'order_freq', 'order_freq_flag'],
      dtype='object')

In [47]:
# rearrange the index of df_ords_merged
# df_ords_merged.reindex(columns = [
#                                   'order_id', 'user_id', 'order_number', 'orders_day_of_week', 'busiest_day', 'busiest_days',
#                                   'add_to_cart_order', 'order_hour_of_day', 'busiest_period_of_day', 
#                                   'max_order', 'loyalty_flag', 'is_new_customer',
#                                   'days_since_prior_order', 'order_freq', 'order_freq_flag', 'reordered', 'avg_price', 'spending_flag',
#                                   'product_id', 'product_name', 'aisle_id', 'department_id', 'prices', 'price_range'
#                                   ])

# my machine ran out of memory at this point so it will be implemented in the next submission

## <a id='toc1_3_'></a>[III. Data Export](#toc0_)

In [48]:
# export df_ords_merged in .pkl format
df_ords_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_3.pkl'))